### ISSUE TITLE
中文不居中

### ISSUE THREAD

**Issue by maplelover** [adobe-fonts/source-han-sans#187]
![861511852598_ pic_hd](https://user-images.githubusercontent.com/6849094/33307214-af8bd144-d450-11e7-8f2d-2ef23f761cdf.jpg)

请教一下：如何处理中文不居中问题

TextView.setIncludeFontPadding(false)

**Comment by Explorer09** [comment]
What do you mean when you say the text is not centered? And can you tell us what software you are using?

**Comment by maplelover** [comment]
This is a android application demo. 

We set the source -han- sans as the system font. When we develop the textview, the top margin and bottom margin are not the same. The red areas in the image show the margins. The text size are 92px in the image, the top margin is 29px, the bottom margin is 21px. The words are not at the center of the text area. 

We also have a problem is the text in android have margins.  But designers mark in photoshop is without margins. So how to solve the problem?


Thanks!

**Comment by Explorer09** [comment]
I don't think this is the font's bug, since your Android textview widget might have the line height properties or margins that affects the display. I'm unfamiliar with Android application development, but could you also test if the same line-height problem also occurs if you are using a different font? Try Droid Sans Fallback, for example.

**Comment by maplelover** [comment]
![1311511873455](https://user-images.githubusercontent.com/6849094/33321012-051a931a-d47f-11e7-94d8-f600a137902b.png)

The two TextView same config except font. The font of bottom TextView is Roboto-Light.ttf.
There is another question for me is that why the the vertical padding of source -han- sans so large.
Looking forward to your reply.

**Comment by Explorer09** [comment]
Roboto Light does not contain CJK characters. And I'm not sure how you get the large vertical padding comparing with your first screenshot.

**Comment by maplelover** [comment]
The TextView widget on android has the method below.

TextView.setIncludeFontPadding(false)

The first screenshot is the result of removing font padding. Don’t you know where the large padding?

**Comment by kenlunde** [comment]
Keep in mind that Android OS's line-layout is still maturing, and that font-fallback plays a big role in how it behaves.

When a font is explicitly specified, such as _Source Han Sans_ or _Roboto_, that font's vertical metrics, meaning the padding that you see above and below the text, is forcibly applied to all fonts in the font-fallback chain. This sort of makes sense, in that the position of each line is predictable. When you specified _Roboto_, its vertical metrics were applied _Noto Sans CJK_, which became referenced because _Roboto_ does not include any glyphs for ideographs.

If you were to specify _Noto Sans CJK_ instead of _Source Han Sans_, you would get the same results, because they share the same vertical metrics. However, the reason why the padding is so huge is because the vertical metrics that are carefully specified in the font (via 'hhea' and 'OS/2' table settings) are being completely ignored, and the top and bottom of the font's bounding box are being used instead. These fonts include vertical forms of U+2E3A ⸺ TWO-EM DASH and U+2E3B ⸻ THREE-EM DASH, along with U+3031 and U+3032 that are two-em tall. As you can see in your examples, the padding is effectively one-em on the top and bottom, with an overall three-em height.

In other words, this is a line-layout problem, not a font problem.

**Comment by maplelover** [comment]
Can you give me the link to download Noto Sans CJK?

**Comment by DanteAndroid** [comment]
Any workaround?
有什么解决办法吗？
